<h1 style="text-align: center;">EDA</h1>


### Checking the right path for the data

In [1]:
from pathlib import Path

# Go one level up from src/ → DDI-RAG
PROJECT_ROOT = Path().resolve().parent

RAW_FOLDER = PROJECT_ROOT / "data" / "raw"
PROCESSED_FOLDER = PROJECT_ROOT / "data" / "processed"

print("Project root:", PROJECT_ROOT)
print("Raw folder exists:", RAW_FOLDER.exists())

Project root: C:\Users\bc0683\ddi-rag
Raw folder exists: True


## Understadning toatl number of drugs appeared with generic names in the datased

In [3]:
import ijson
from pathlib import Path

PROJECT_ROOT = Path().resolve().parent
RAW_FOLDER = PROJECT_ROOT / "data" / "raw"

all_drug_names = []

for file in RAW_FOLDER.glob("drug-label-*.json"):
    print(f"Processing {file.name}...")
    
    with open(file, "rb") as f:
        for drug in ijson.items(f, "results.item"):
            generic_names = drug.get("openfda", {}).get("generic_name")
            
            if generic_names:
                all_drug_names.append(generic_names[0].lower())

print("\nTotal drug entries collected:", len(all_drug_names))

Processing drug-label-0001-of-0013.json...
Processing drug-label-0002-of-0013.json...
Processing drug-label-0003-of-0013.json...
Processing drug-label-0004-of-0013.json...
Processing drug-label-0005-of-0013.json...
Processing drug-label-0006-of-0013.json...
Processing drug-label-0007-of-0013.json...
Processing drug-label-0008-of-0013.json...
Processing drug-label-0009-of-0013.json...
Processing drug-label-0010-of-0013.json...
Processing drug-label-0011-of-0013.json...
Processing drug-label-0012-of-0013.json...
Processing drug-label-0013-of-0013.json...

Total drug entries collected: 82945


# Understading total number of drugs present in the datset
## . how many unique data labels are present.

## . how many drugs with interaction infomation in available


In [4]:
import ijson

total_drugs = 0
drugs_with_interactions = 0
unique_names = set()

for file in RAW_FOLDER.glob("drug-label-*.json"):
    print(f"Processing {file.name}...")
    
    with open(file, "rb") as f:
        for drug in ijson.items(f, "results.item"):
            total_drugs += 1
            
            generic_names = drug.get("openfda", {}).get("generic_name")
            interactions = drug.get("drug_interactions")
            
            if generic_names:
                unique_names.add(generic_names[0].lower())
            
            if interactions:
                drugs_with_interactions += 1

print("\n===== SUMMARY =====")
print("Total drug entries:", total_drugs)
print("Unique drug names:", len(unique_names))
print("Drugs with interaction info:", drugs_with_interactions)

Processing drug-label-0001-of-0013.json...
Processing drug-label-0002-of-0013.json...
Processing drug-label-0003-of-0013.json...
Processing drug-label-0004-of-0013.json...
Processing drug-label-0005-of-0013.json...
Processing drug-label-0006-of-0013.json...
Processing drug-label-0007-of-0013.json...
Processing drug-label-0008-of-0013.json...
Processing drug-label-0009-of-0013.json...
Processing drug-label-0010-of-0013.json...
Processing drug-label-0011-of-0013.json...
Processing drug-label-0012-of-0013.json...
Processing drug-label-0013-of-0013.json...

===== SUMMARY =====
Total drug entries: 254983
Unique drug names: 14112
Drugs with interaction info: 66695


### . from the two cells we can observe that there are a totoal of *254K enteries in thh dataset*, whith a *83K generic names* , *unique named druges with around 14K* and *drugs with interactions around 66k*  

| Metric  | What It Counts                |
| ------- | ----------------------------- |
| 254,983 | All label records             |
| 82,945  | Records with generic names    |
| 66,695  | Records with interaction data |
| 14,112  | Unique drug names             |


In [5]:
all_drug_names[:50]

['amoxicillin and clavulanate potassium',
 'emtricitabine and tenofovir disoproxil fumarate',
 'triple antibiotic ointment',
 'acetaminophen, aspirin, and caffeine',
 'titanium dioxide, zinc oxide',
 'buspirone hydrochloride',
 'olmesartan medoxomil / amlodipine besylate / hydrochlorothiazide',
 'linagliptin and metformin hydrochloride',
 'angelica archangelica, drosera, grindelia, lamium album, pinus sylvestris, rubus fruticosus, senega officinalis',
 'pertuzumab, trastuzumab, and hyaluronidase-zzxf',
 'guaifenesin and dextromethorphan hbr',
 'octinoxate, zinc oxide',
 'octinoxate, zinc oxide',
 'dextroamphetamine sulfate, dextroamphetamine saccharate, amphetamine sulfate and amphetamine aspartate',
 'isopropyl alcohol',
 'titanium dioxide, zinc oxide',
 'dextroamphetamine saccharate, amphetamine aspartate, dextroamphetamine sulfate and amphetamine sulfate',
 'water',
 'diclofenac sodium and misoprostol',
 'potassium phosphate, monobasic potassium phosphate, dibasic',
 'estradiol',
 '

In [6]:
from collections import Counter

drug_counts = Counter(all_drug_names)

drug_counts.most_common(20)

[('zinc oxide', 1812),
 ('alcohol', 1419),
 ('acetaminophen', 965),
 ('ibuprofen', 916),
 ('salicylic acid', 910),
 ('menthol', 851),
 ('benzalkonium chloride', 794),
 ('sodium fluoride', 784),
 ('titanium dioxide, zinc oxide', 640),
 ('avobenzone, homosalate, octisalate, octocrylene', 595),
 ('nicotine polacrilex', 520),
 ('ethyl alcohol', 517),
 ('isopropyl alcohol', 514),
 ('oxygen', 449),
 ('lidocaine', 430),
 ('calcium carbonate', 422),
 ('gabapentin', 391),
 ('benzocaine', 385),
 ('aspirin', 370),
 ('hydrocortisone', 349)]

In [7]:
list(unique_names)[:20]

['bicisate dihydrochloride',
 'menthol, methylsalicylate',
 'acetaminophen, dextromethorphan hydrobromide and doxylamine succinate',
 'flu nosode 2021-2022',
 'non nano zinc oxide',
 'calcitriol',
 'fourth chakra (heart)',
 'loperamide hydrochloride and simethicone',
 'sarecycline hydrochloride',
 'atidarsagene autotemcel',
 '.beta.-carotene, ascorbic acid, cholecalciferol, .alpha.-tocopherol acetate, dl-, thiamine mononitrate, riboflavin, niacinamide, pyridoxine hydrochloride, folic acid, 5-methyltetrahydrofolic acid, calcium formate, ferrous asparto glycinate, cyanocobalamin, biotin, potassium iodide, magnesium oxide, zinc oxide and cupric oxide',
 'sodium chloride, potassium chloride, calcium chloride, magnesium chloride, sodium acetate, and sodium citrate',
 'sotalol hydrochloride',
 'beeswax spf 15 broad spectrum lip balm',
 'biotitum',
 'adapalene and benzoyl peroxide gel, 0.1%/2.5%',
 'benzocaine and zinc chloride',
 'renes cuprum',
 'calendula officinalis, caladium seguinum, le

In [8]:
files = list(RAW_FOLDER.glob("drug-label-*.json"))[:2]

In [10]:

import pandas as pd 
df = pd.DataFrame({"drug_name": all_drug_names})

print(df.head())
print("Total rows:", len(df))

                                         drug_name
0            amoxicillin and clavulanate potassium
1  emtricitabine and tenofovir disoproxil fumarate
2                       triple antibiotic ointment
3             acetaminophen, aspirin, and caffeine
4                     titanium dioxide, zinc oxide
Total rows: 82945


In [11]:
df["drug_name"].value_counts().head(20)

drug_name
zinc oxide                                         1812
alcohol                                            1419
acetaminophen                                       965
ibuprofen                                           916
salicylic acid                                      910
menthol                                             851
benzalkonium chloride                               794
sodium fluoride                                     784
titanium dioxide, zinc oxide                        640
avobenzone, homosalate, octisalate, octocrylene     595
nicotine polacrilex                                 520
ethyl alcohol                                       517
isopropyl alcohol                                   514
oxygen                                              449
lidocaine                                           430
calcium carbonate                                   422
gabapentin                                          391
benzocaine                            

In [12]:
count_interactions = 0

for file in RAW_FOLDER.glob("drug-label-*.json"):
    with open(file, "rb") as f:
        for drug in ijson.items(f, "results.item"):
            if drug.get("drug_interactions"):
                count_interactions += 1

print("Total entries with interactions:", count_interactions)

Total entries with interactions: 66695
